Initial rails application template

tekin opened the pull request:
A first-pass rails app template for creating and configuring a bare Rails application with common GOV.UK features. This should save a fair amount of time when starting a fresh Rails application on GOV.UK.

As outlined in the [README](https://github.com/alphagov/govuk-rails-app-template/blob/044e15b07201fe1e77d88a6212c2d4c5b1c7a1c0/README.md#how-to-use), a new app can be generated as follows:

``` shell
rails new APP_NAME --skip-test-unit --skip-spring -m govuk-rails-app-template/template.rb
```

This will generate a new rails application, and also do the following:
- Install and configure rspec for testing, skipping test-unit
- Install and configure logstasher for JSON-formatted log output in production
- Install and configure simple-cov for code coverage reporting
- Add a `.ruby-version` file
- Add a skeleton `README.md` based on the [styleguide](https://github.com/alphagov/styleguides/blob/master/use-of-READMEs.md)
- Add a `LICENSE` file
- Add jenkins scripts for master and branch builds
- Add a `/healthcheck` endpoint, including test

This makes use of the [Rails application templating](http://guides.rubyonrails.org/rails_application_templates.html) functionality, which is a wrapper around [thor](http://www.rubydoc.info/github/erikhuda/thor/master/Thor/Actions).

tekin committed:
Configure JSON-formatted loggin

tekin committed:
Add a .ruby-version file

tekin committed:
Set up README.md and LINCENSE files

tekin committed:
Read logstash initializer from a file

tekin committed:
Add a healthcheck route

Will add some specs once rspec is configured

tekin committed:
Jenkins master and branch build scripts

tekin committed:
Install rspec

Updates the README to reflect that test/unit should be skipped.

tekin committed:
Delete test/unit folder

tekin committed:
Add spec for healthcheck endpoint

tekin committed:
Skip spring

Is a bit flakey - doesn't appear to play nicely with rspec

tekin committed:
Generate app with atomic commits

This will give the repo a richer history

tekin committed:
Install and configure code coverage reporting

tekin commented:
Here is an example application generated using the above template: https://github.com/tekin/govuk-test-app/commits/master

alext commented on template.rb:
```diff
@@ -0,0 +1,83 @@
+
+# Boilerplate jenkins scripts
+copy_file 'templates/jenkins.sh', 'jenkins.sh'
+template  'templates/jenkins_branches.sh.erb', 'jenkins_branches.sh'
```
Will this make the scripts executable?  They need to be.

alext commented on template.rb:
```diff
@@ -0,0 +1,83 @@
+  gem 'simplecov-rcov', :require => false
+end
+
+inject_into_file 'spec/rails_helper.rb', after: "require 'rspec/rails'\n" do <<-'RUBY'
```
This should be at the very top of `spec_helper.rb`.  For simplecov to work correctly it needs to be loaded before any application code etc.

alext commented on template.rb:
```diff
@@ -0,0 +1,83 @@
+
+# Setup rspec
+gem_group :development, :test do
+  gem 'rspec-rails'
```
It would be good to pin the version here to `'~> 3.3'`.  They follow semantic versioning - http://www.relishapp.com/rspec/docs/versioningpolicy.

alext commented on template.rb:
```diff
@@ -0,0 +1,83 @@
+run 'bundle install'
+git :init
+git add: "."
+git commit: "-a -m 'Bare Rails application'"
```
It's probably worth including a link to this repo in the commit message to given an indication how this was generated.

alext commented:
A few comments, but otherwise this is a really good thing.  Thanks for getting this setup.

tekin committed:
Lock rspec-rails version

tekin committed:
Load simplecov early

simplecov needs to be loaded before the application code is loaded, otherwise
it will not provide accurate coverage results.

tekin committed:
Ensure jenkins scripts are executable

tekin committed:
Include app generator link in commit message

tekin commented:
Thanks for the comments, I think I've addressed them all.

alext commented:
On Gem versioning, it should be enough to pin rspec to `~> 3.3` as they follow semver.

Another point I noticed is that we're not pinning the version of simplecov.  Would be good to do that as well.

Oh, and also logstasher.

alext commented on README.md:
````diff
@@ -6,7 +6,7 @@ GOV.UK stack.
 
 ```shell
-rails new APP_NAME -m http://raw.github.com/govuk-rails-app-template/master/template.rb
+rails new APP_NAME --skip-test-unit --skip-spring -m govuk-rails-app-template/template.rb
````
This command adds several gems that we may not want, such as turbolinks and coffee-rails.  We should probably suggest using `--no-javascript` here to omit these.  That will leave out jquery as well, but most apps that use javascript, will be either using govuk_admin_template, or slimmer to provide jquery.

Another option would be to replace the Gemfile with a fully curated one.

alext replied:
It's also worth prefixing this command with `RBENV_VERSION=2.2.2`, otherwise it'll run under the default ruby, which is typically the system ruby.

tekin replied:
Updated with `--no-javascript` and RBENV prefix. 

I think we'll ultimately end up with either separate templates for the different types of app, or the template will become interactive, e.g.

``` ruby
if yes?("is this a frontend application?")
  # setup with frontend stuff, e.g. slimmer, etc
elsif yes?("is this an admin application?")
  # setup with admin stuff, e.g. govuk_admin_template, gds-sso, etc
else
  # it's something else, like authenticating-proxy
end
```

tekin committed:
Update "How to use" in README

We will almost certainly be using either govuk-admin-template or
govuk-frontend-toolkit/slimmer for javascript, etc.

tekin committed:
Lock gem versions to known good configuration
